### PR TITLE
keep placed canaries aligned in raft store

### LINKED ISSUE #6864: Nomad kills off healthy allocations on deploy when new allocations fail
### Nomad version

```
$ nomad -v
Nomad v0.10.2 (0d2d6e3dc5a171c21f8f31fa117c8a765eb4fc02)
```

### Operating system and Environment details

Amazon Linux 2 running on AWS.

### Issue

When we are deploying using the Blue/Green canary deployment model we have 4 allocations running (healthy), and we try to boot up 4 new allocations and they fail, Nomad retries that, and somehow Nomad kills off the healthy and running allocations leaving us with no healthy allocations running:

screenie_1576449035_369936

### Reproduction steps

Deploy a healthy job, after that deploy a failing job, observe what happens.

### Job file

```hcl
job "billetto-production-web" {
  datacenters = ["dc1"]
  type = "service"

  update {
    max_parallel = 4
    min_healthy_time = "10s"
    healthy_deadline = "3m"
    progress_deadline = "10m"
    auto_revert = false
    auto_promote = false
    canary = 4
  }

  migrate {
    max_parallel = 3
    health_check = "checks"
    min_healthy_time = "10s"
    healthy_deadline = "5m"
  }

  group "group" {
    count = 4

    restart {
      attempts = 0
      interval = "30m"
      delay = "15s"
      mode = "fail"
    }

    ephemeral_disk {
      size = 300
    }

    task "nginx" {
      driver = "docker"

      config {
        image = "billetto/billetto-nginx:42.42.42"

        port_map {
          nginx = 80
        }

        dns_servers = ["172.17.0.1"]
      }

      resources {
        cpu    = 200
        memory = 170
        network {
          mbits = 50
          port "nginx" {}
        }
      }

      service {
        name = "billetto-production-nginx"
        tags = []
        port = "nginx"

        check {
          name     = "billetto-production-web nginx healthcheck"
          type     = "http"
          protocol = "http"
          path     = "/debug/healthcheck"
          interval = "5s"
          timeout  = "3s"
        }
      }
    }

    task "rails" {
      driver = "docker"

      config {
        image = "billetto/billetto:42.42.42"

        command = "bin/bundle"
        args = ["exec", "unicorn", "-c", "config/unicorn.rb", "--no-default-middleware"]

        port_map {
          web = 3000
        }

        dns_servers = ["172.17.0.1"]
      }

      resources {
        cpu    = 750
        memory = 1650
        network {
          mbits = 50
          port "web" {}
        }
      }

      service {
        name = "billetto-production-web"
        tags = []
        port = "web"

        check {
          name     = "billetto-production-web rails healthcheck"
          type     = "http"
          protocol = "http"
          path     = "/debug/healthcheck"
          interval = "5s"
          timeout  = "3s"
        }
      }
    }
  }
}
```

### LINKED ISSUE #6936: Failed canary allocations are rescheduled as non canary allocations
### Nomad version
0.10.3-dev

### Operating system and Environment details
ubuntu 19.10

### Issue

[Rescheduling during deployments doc](https://www.nomadproject.io/docs/job-specification/reschedule.html#rescheduling-during-deployments) claims that 
> The update stanza controls rolling updates and canary deployments. A task group's reschedule stanza does not take affect during a deployment

However, with the reproduction steps & file listed, a failed set of canary allocations get rescheduled according to the reschedule stanza, but _do not_ register as canary allocations. This causes the previous deploys healthy tasks to be marked as stop during reconciliation.

### Reproduction steps

using a local cluster with 3 server nodes, 2 client nodes

```
# run first job
nomad job run repro.hcl

# allow deploy to become healthy

nomad job run repro-fail.hcl

# observe rescheduled eval eventually kill off healthy allocations
```
Failed canaries, future reschedule occurring shortly
```
ID            = failing-nomad-test01
Name          = failing-nomad-test01
Submit Date   = 2020-01-13T15:35:51-05:00
Type          = service
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = running
Periodic      = false
Parameterized = false

Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost
group       0       0         4        4       0         0

Future Rescheduling Attempts
Task Group  Eval ID   Eval Time
group       d8ddc12d  2s from now

Latest Deployment
ID          = 86d8d83f
Status      = running
Description = Deployment is running but requires manual promotion

Deployed
Task Group  Promoted  Desired  Canaries  Placed  Healthy  Unhealthy  Progress Deadline
group       false     4        4         4       0        4          2020-01-13T15:45:51-05:00

Allocations
ID        Node ID   Task Group  Version  Desired  Status   Created  Modified
78bc2bb6  303bd631  group       1        run      failed   30s ago  26s ago
164b81e5  303bd631  group       1        run      failed   30s ago  26s ago
203a01c6  303bd631  group       1        run      failed   30s ago  26s ago
497ae378  303bd631  group       1        run      failed   30s ago  26s ago
18df0793  27ce5ad0  group       0        run      running  55s ago  40s ago
2949e33b  27ce5ad0  group       0        run      running  55s ago  36s ago
0e191e2a  27ce5ad0  group       0        run      running  55s ago  38s ago
e5f6b54c  27ce5ad0  group       0        run      running  55s ago  37s ago

```

Newly placed allocations cause original healthy allocs to be marked as stop
```
ID            = failing-nomad-test01
Name          = failing-nomad-test01
Submit Date   = 2020-01-13T15:35:51-05:00
Type          = service
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = pending
Periodic      = false
Parameterized = false

Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost
group       0       0         1        8       3         0

Future Rescheduling Attempts
Task Group  Eval ID   Eval Time
group       289de16b  58s from now

Latest Deployment
ID          = 86d8d83f
Status      = running
Description = Deployment is running but requires manual promotion

Deployed
Task Group  Promoted  Desired  Canaries  Placed  Healthy  Unhealthy  Progress Deadline
group       false     4        4         8       0        8          2020-01-13T15:45:51-05:00

Allocations
ID        Node ID   Task Group  Version  Desired  Status    Created   Modified
5824276a  303bd631  group       1        run      failed    4s ago    0s ago
19cb409a  303bd631  group       1        run      failed    4s ago    0s ago
8c70b8cc  303bd631  group       1        run      failed    4s ago    0s ago
8c5e464e  27ce5ad0  group       1        run      failed    4s ago    0s ago
497ae378  303bd631  group       1        stop     failed    36s ago   4s ago
78bc2bb6  303bd631  group       1        stop     failed    36s ago   4s ago
164b81e5  303bd631  group       1        stop     failed    36s ago   4s ago
203a01c6  303bd631  group       1        stop     failed    36s ago   4s ago
2949e33b  27ce5ad0  group       0        stop     running   1m1s ago  0s ago
0e191e2a  27ce5ad0  group       0        stop     complete  1m1s ago  0s ago
18df0793  27ce5ad0  group       0        stop     complete  1m1s ago  0s ago
e5f6b54c  27ce5ad0  group       0        stop     complete  1m1s ago  0s ago

```

### Job file (if appropriate)

repro.hcl
```
job "failing-nomad-test01" {
  datacenters = ["dc1"]
  type        = "service"

  update {
    health_check      = "checks"
    max_parallel      = 4
    min_healthy_time  = "10s"
    healthy_deadline  = "3m"
    progress_deadline = "10m"
    auto_revert       = false
    auto_promote      = false
    canary            = 4
  }

  migrate {
    max_parallel     = 3
    health_check     = "checks"
    min_healthy_time = "10s"
    healthy_deadline = "5m"
  }

  group "group" {
    count = 4

    restart {
      attempts = 0
      interval = "30m"
      delay    = "15s"
      mode     = "fail"
    }

    ephemeral_disk {
      size = 300
    }

    task "rails" {
      driver = "docker"

      env {
        RAILS_ENV                = "production"
        RAILS_SERVE_STATIC_FILES = "1"
        RAILS_LOG_TO_STDOUT      = "1"
        TEST                     = "2020-01-09 01:23:18 +0000"
      }

      config {
        image = "kaspergrubbe/diceapp:0.0.6"

        command = "bundle"
        args    = ["exec", "unicorn", "-c", "/app/config/unicorn.rb"]

        port_map {
          web = 8080
        }
      }

      resources {
        cpu    = 750
        memory = 250

        network {
          mbits = 50
          port  "web" {}
        }
      }

      service {
        name = "failing-nomad-test00"
        tags = []
        port = "web"

        check {
          name     = "failing-nomad-test00 healthcheck"
          type     = "http"
          protocol = "http"
          path     = "/"
          interval = "5s"
          timeout  = "3s"
        }
      }
    }
  }
}
```

repro-fail.hcl
```
job "failing-nomad-test01" {
  datacenters = ["dc1"]
  type        = "service"

  update {
    health_check      = "checks"
    max_parallel      = 4
    min_healthy_time  = "10s"
    healthy_deadline  = "3m"
    progress_deadline = "10m"
    auto_revert       = false
    auto_promote      = false
    canary            = 4
  }

  migrate {
    max_parallel     = 3
    health_check     = "checks"
    min_healthy_time = "10s"
    healthy_deadline = "5m"
  }

  group "group" {
    count = 4

    restart {
      attempts = 0
      interval = "30m"
      delay    = "15s"
      mode     = "fail"
    }

    ephemeral_disk {
      size = 300
    }

    task "rails" {
      driver = "docker"

      env {
        RAILS_ENV                = "production"
        RAILS_SERVE_STATIC_FILES = "1"
        RAILS_LOG_TO_STDOUT      = "1"
        TEST                     = "2020-01-09 01:23:18 +0000"
      }

      config {
        image = "kaspergrubbe/diceapp:0.0.6"

        command = "false"

        port_map {
          web = 8080
        }

        dns_servers = ["172.17.0.1"]
      }

      resources {
        cpu    = 750
        memory = 250

        network {
          mbits = 50
          port  "web" {}
        }
      }

      service {
        name = "failing-nomad-test00"
        tags = []
        port = "web"

        check {
          name     = "failing-nomad-test00 healthcheck"
          type     = "http"
          protocol = "http"
          path     = "/"
          interval = "5s"
          timeout  = "3s"
        }
      }
    }
  }
}

```

originally reported issue #6864

### PR DESCRIPTION
**Pull request by drewbailey**
Nomad state store must be modified through raft; `DeploymentState.PlacedCanaries` is currently [modified in-place](https://github.com/hashicorp/nomad/blob/master/scheduler/generic_sched.go#L532-L534) which masked a bug in single node clusters, but made evaluations in other nodes unaware of the true state of Canaries during a deployment.

This change reconciles the state of `PlacedCanaries` when committing deployment/alloc updates to raft instead of modifying local state of a specific scheduler.

fixes #6936 
fixes #6864

### CODE REVIEW

File: nomad/state/state_store.go [hunk 1]
```diff
@@ -3442,6 +3442,21 @@ func (s *StateStore) updateDeploymentWithAlloc(index uint64, alloc, existing *st
 	state.HealthyAllocs += healthy
 	state.UnhealthyAllocs += unhealthy
 
+	// Ensure PlacedCanaries accurately reflects the alloc canary status
+	if alloc.DeploymentStatus != nil {
+		if alloc.DeploymentStatus.Canary {
+			found := false
+			for _, canary := range state.PlacedCanaries {
+				if alloc.ID == canary {
+					found = true
```
Comment (notnoop) [inline nomad/state/state_store.go]: You can break once you find the element
```suggestion
					found = true
					break
```

Comment (drewbailey) [reply nomad/state/state_store.go]: :sweat_smile: woops thanks

File: scheduler/generic_sched_test.go [hunk 2]
```diff
@@ -2029,17 +2029,29 @@ func TestServiceSched_JobModify_Canaries(t *testing.T) {
...
+	ws := memdb.NewWatchSet()
+
+	// Grab the latest state
+	deploy, err := h.State.DeploymentByID(ws, plan.Deployment.ID)
+	assert.NoError(t, err)
+
+	state, ok := deploy.TaskGroups[job.TaskGroups[0].Name]
+	assert.True(t, ok)
+
 	if state.DesiredTotal != 10 && state.DesiredCanaries != desiredUpdates {
```
Comment (notnoop) [inline scheduler/generic_sched_test.go]: I'd suggest using `require` here - unlike `assert`, `require` terminate on failure and avoids panics/null deference (e.g. consider when `ok` is false in the test.

Also, you can replace this conditional and the one below with `require.Equal` too.

File: nomad/state/state_store_test.go [hunk 3]
```diff
@@ -6382,6 +6382,39 @@ func TestStateStore_UpsertDeploymentAllocHealth_BadAlloc_Nonexistent(t *testing.
...
+	}
+	require.NoError(t, state.UpsertAllocs(4, []*structs.Allocation{a}))
+
+	// Pull the deployment from state
+	ws := memdb.NewWatchSet()
+	deploy, err := state.DeploymentByID(ws, d1.ID)
+	require.NoError(t, err)
+
+	// Ensure that PlacedCanaries is accurate
+	require.Equal(t, 1, len(deploy.TaskGroups[job.TaskGroups[0].Name].PlacedCanaries))
```
Comment (notnoop) [inline nomad/state/state_store_test.go]: Nice clear test.  I'd suggest adding a couple of more conditions: when allocs for multiple deployments exist (only relevant deployment for associated with alloc is updated): when canary is false (no update).

File: nomad/state/state_store.go [hunk 4]
```diff
@@ -3442,6 +3442,20 @@ func (s *StateStore) updateDeploymentWithAlloc(index uint64, alloc, existing *st
 	state.HealthyAllocs += healthy
 	state.UnhealthyAllocs += unhealthy
 
+	// Ensure PlacedCanaries accurately reflects the alloc canary status
+	if alloc.DeploymentStatus != nil && alloc.DeploymentStatus.Canary {
```
Comment (notnoop) [inline nomad/state/state_store.go]: Actually, do we need to update `UpdateDeploymentPromotion` so that canaries that get promoted are removed from PlacedCanaries?

Comment (drewbailey) [reply nomad/state/state_store.go]: I don't think so, since `UpdateDeploymentPromotion` deals with promoting the deployment/canaries I think that the values for Placed shouldn't ever be decremented. If you want to see the deployment status of a previous deployment the value for Placed should be the total that were placed

Comment (notnoop) [reply nomad/state/state_store.go]: Umm.. I'm a bit concerned about the reconciler logic.  Currently, `handleGroupCanaries` returns `canaries`, by checking `PlacedCanaries` exclusively.  If a canary got promoted, how do we ensure that scheduler no longer treats it as one.

Comment (drewbailey) [reply nomad/state/state_store.go]: I believe (and can followup / confirm) Canary promotion only happens when a deployment is promoted. If there is an old deployment with non-promoted canaries, or the deployment is failed, it will use PlacedCanaries that haven't been promoted to be stopped.
```
	// Cancel any non-promoted canaries from the older deployment
	if a.oldDeployment != nil {
		for _, s := range a.oldDeployment.TaskGroups {
			if !s.Promoted {
				stop = append(stop, s.PlacedCanaries...)
			}
		}
	}

	// Cancel any non-promoted canaries from a failed deployment
	if a.deployment != nil && a.deployment.Status == structs.DeploymentStatusFailed {
		for _, s := range a.deployment.TaskGroups {
			if !s.Promoted {
				stop = append(stop, s.PlacedCanaries...)
			}
		}
	}
```
And then after the deployment is promoted, the promotion eval includes the promoted deployment canaries as canaries, but then further down `canaryState` will be false
```
	canaryState := dstate != nil && dstate.DesiredCanaries != 0 && !dstate.Promoted
```
Which is used in computeStop to stop the previous version non canary allocs. I think after all of this, the deploy is marked as successful, and even though placed canaries stays populated, we check deployment status and no new canaries are ever placed until a new deployment

File: scheduler/generic_sched.go [hunk 5]
```diff
@@ -529,10 +529,6 @@ func (s *GenericScheduler) computePlacements(destructive, place []placementResul
 				// If we are placing a canary and we found a match, add the canary
 				// to the deployment state object and mark it as a canary.
 				if missing.Canary() && s.deployment != nil {
-					if state, ok := s.deployment.TaskGroups[tg.Name]; ok {
-						state.PlacedCanaries = append(state.PlacedCanaries, alloc.ID)
-					}
```
Comment (schmichael) [inline scheduler/generic_sched.go]: I'm unclear why an inplace update is inappropriate or the source of a bug here. It's hard to follow this flow, but `s.deployment` is set by the reconciler which creates a *copy* of the deployment:

https://github.com/hashicorp/nomad/blob/f5441e69d4c991b77318bb661b31ab6acb2c7ea7/scheduler/reconcile.go#L170

This seems like an intentional decision to make deployments safe for inplace updates, and it appears `s.deployment == s.Plan.Deployment` which means it will get submitted along with the plan.

If that's not the case I'd love to see a little more cleanup or commenting in the code to help future developers. Maybe an explicit comment on the `s.deployment` field about its mutability, and removing the `deployment.Copy()` if it is immutable.

Comment (drewbailey) [reply scheduler/generic_sched.go]: So there is definitely some misdirection going on with all these `deployment`s I'm going to outline my understanding of them and then see about commenting or cleaning up.

>  but s.deployment is set by the reconciler which creates a copy of the deployment

The scheduler deployment is actually not set by the reconciler, the reconciler copies the scheduler deployment to its own `deployment` and uses it as a readonly reference to the current state of the deployment and returns a `results.deployment` and `results.deploymentUpdates`

https://github.com/hashicorp/nomad/blob/master/scheduler/generic_sched.go#L355-L357

Currently the reconciler will only return a `results.deployment` if the deployment is newly created, otherwise it will just be `deploymentUpdates`.

So the following logic is mutating state that never gets written to the raft log because only changes on `s.plan` are ultimately submitted, in a single node cluster the state will remain locally modified and show the correct number of canaries but in multi node clusters a different worker won't have these changes
```go
if state, ok := s.deployment.TaskGroups[tg.Name]; ok {
	state.PlacedCanaries = append(state.PlacedCanaries, alloc.ID)
}
```

**Comment by drewbailey** [comment]
Leaving reproduction steps for testing purposes

Setup and run a local dev cluster (https://github.com/hashicorp/nomad/blob/master/dev/cluster/cluster.sh)

<details>
<summary> run repo.hc.</summary>


```hcl
job "failing-nomad-test01" {
  datacenters = ["dc1"]
  type        = "service"

  constraint {
    attribute = "${meta.tag}"
    value     = "foo"
  }

  update {
    health_check      = "checks"
    max_parallel      = 4
    min_healthy_time  = "10s"
    healthy_deadline  = "3m"
    progress_deadline = "10m"
    auto_revert       = false
    auto_promote      = false
    canary            = 4
  }

  migrate {
    max_parallel     = 3
    health_check     = "checks"
    min_healthy_time = "10s"
    healthy_deadline = "5m"
  }

  group "group" {
    count = 4

    restart {
      attempts = 0
      interval = "30m"
      delay    = "15s"
      mode     = "fail"
    }

    ephemeral_disk {
      size = 300
    }

    task "rails" {
      driver = "docker"

      env {
        RAILS_ENV                = "production"
        RAILS_SERVE_STATIC_FILES = "1"
        RAILS_LOG_TO_STDOUT      = "1"
        TEST                     = "2020-01-09 01:23:18 +0000"
        version                  = "129"
      }

      config {
        image = "kaspergrubbe/diceapp:0.0.6"

        command = "bundle"
        args    = ["exec", "unicorn", "-c", "/app/config/unicorn.rb"]

        port_map {
          web = 8080
        }
      }

      resources {
        cpu    = 750
        memory = 250

        network {
          mbits = 50
          port  "web" {}
        }
      }

      service {
        name = "failing-nomad-test00"
        tags = []
        port = "web"

        check {
          name     = "failing-nomad-test00 healthcheck"
          type     = "http"
          protocol = "http"
          path     = "/"
          interval = "5s"
          timeout  = "3s"
        }
      }
    }
  }
}
```
</details>


wait for initial deployment to become healthy `nomad job status -verbose failing-`
```
Deployed
Task Group  Desired  Placed  Healthy  Unhealthy  Progress Deadline
group       4        4       4        0          2020-02-03T13:41:42-05:00

Allocations
ID                                    Eval ID                               Node ID                               Node Name  Task Group  Version  Desired  Status   Created                    Modified
3c7eb6cb-aea9-055f-96e6-760eb3526d4a  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        run      running  2020-02-03T13:31:22-05:00  2020-02-03T13:31:42-05:00
3f34931c-0e25-bd86-7ca1-1535d945a180  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        run      running  2020-02-03T13:31:22-05:00  2020-02-03T13:31:41-05:00
5099bcaf-336b-99ad-40c7-dcb20d39fd47  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        run      running  2020-02-03T13:31:22-05:00  2020-02-03T13:31:39-05:00
d48180bc-021e-93c0-0569-9812329a9d21  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        run      running  2020-02-03T13:31:22-05:00  2020-02-03T13:31:39-05:00
```

<details>
<summary>start a deploy with failing job `nomad job run repro-fail.hcl</summary>
```hcl
job "failing-nomad-test01" {
  datacenters = ["dc1"]
  type        = "service"

  constraint {
    attribute = "${meta.tag}"
    value     = "foo"
  }

  update {
    health_check      = "checks"
    max_parallel      = 4
    min_healthy_time  = "10s"
    healthy_deadline  = "3m"
    progress_deadline = "10m"
    auto_revert       = false
    auto_promote      = false
    canary            = 4
  }

  migrate {
    max_parallel     = 3
    health_check     = "checks"
    min_healthy_time = "10s"
    healthy_deadline = "5m"
  }

  group "group" {
    count = 4

    restart {
      attempts = 0
      interval = "30m"
      delay    = "15s"
      mode     = "fail"
    }

    ephemeral_disk {
      size = 300
    }

    task "rails" {
      driver = "docker"

      env {
        RAILS_ENV                = "production"
        RAILS_SERVE_STATIC_FILES = "1"
        RAILS_LOG_TO_STDOUT      = "1"
        TEST                     = "2020-01-09 01:23:18 +0000"
      }

      config {
        image = "kaspergrubbe/diceapp:0.0.6"

        command = "false"

        port_map {
          web = 8080
        }
      }

      resources {
        cpu    = 750
        memory = 250

        network {
          mbits = 50
          port  "web" {}
        }
      }

      service {
        name = "failing-nomad-test00"
        tags = []
        port = "web"

        check {
          name     = "failing-nomad-test00 healthcheck"
          type     = "http"
          protocol = "http"
          path     = "/"
          interval = "5s"
          timeout  = "3s"
        }
      }
    }
  }
}

```

</details>

observe 4 failed canaries, and delated rescheduling attempt

```
Future Rescheduling Attempts
Task Group  Reschedule Policy                                     Eval ID                               Eval Time
group       unlimited with exponential delay, max_delay = 1h0m0s  1f3bc3c6-5a29-2e7f-ddda-2ba93a51b8e6  25s from now

Latest Deployment
ID          = 533e4c07-acfc-444a-01ee-ecd93829e4c4
Status      = running
Description = Deployment is running but requires manual promotion

Deployed
Task Group  Promoted  Desired  Canaries  Placed  Healthy  Unhealthy  Progress Deadline
group       false     4        4         4       0        4          2020-02-03T13:42:56-05:00

Allocations
ID                                    Eval ID                               Node ID                               Node Name  Task Group  Version  Desired  Status   Created                    Modified
3a19b38c-e587-3a5c-9bac-c7dc9f80da70  7a78c791-e553-3d4a-f4f1-9426e2b5c3dd  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        run      failed   2020-02-03T13:32:56-05:00  2020-02-03T13:32:59-05:00
e3396874-555d-5651-81e0-99a4ef74826d  7a78c791-e553-3d4a-f4f1-9426e2b5c3dd  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        run      failed   2020-02-03T13:32:56-05:00  2020-02-03T13:32:59-05:00
9f00f827-63ba-da46-8156-c9c5413bf1f4  7a78c791-e553-3d4a-f4f1-9426e2b5c3dd  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        run      failed   2020-02-03T13:32:56-05:00  2020-02-03T13:32:59-05:00
b1a8d187-5edd-688f-e3e2-19b19f5d5e0c  7a78c791-e553-3d4a-f4f1-9426e2b5c3dd  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        run      failed   2020-02-03T13:32:56-05:00  2020-02-03T13:32:59-05:00
3f34931c-0e25-bd86-7ca1-1535d945a180  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        run      running  2020-02-03T13:31:22-05:00  2020-02-03T13:31:41-05:00
5099bcaf-336b-99ad-40c7-dcb20d39fd47  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        run      running  2020-02-03T13:31:22-05:00  2020-02-03T13:31:39-05:00
d48180bc-021e-93c0-0569-9812329a9d21  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        run      running  2020-02-03T13:31:22-05:00  2020-02-03T13:31:39-05:00
3c7eb6cb-aea9-055f-96e6-760eb3526d4a  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        run      running  2020-02-03T13:31:22-05:00  2020-02-03T13:31:42-05:00
```

Once rescheduled eval runs, it should kill off the 4 healthy previous allocs instead of placing 4 new canaries

```
Latest Deployment
ID          = 533e4c07-acfc-444a-01ee-ecd93829e4c4
Status      = running
Description = Deployment is running but requires manual promotion

Deployed
Task Group  Promoted  Desired  Canaries  Placed  Healthy  Unhealthy  Progress Deadline
group       false     4        4         8       0        8          2020-02-03T13:42:56-05:00

Allocations
ID                                    Eval ID                               Node ID                               Node Name  Task Group  Version  Desired  Status    Created                    Modified
9f2a9488-0d81-ebb2-be52-c0e00d541ae9  1f3bc3c6-5a29-2e7f-ddda-2ba93a51b8e6  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        run      failed    2020-02-03T13:33:27-05:00  2020-02-03T13:33:30-05:00
b9b4bd00-510a-9846-9cac-60f34eeb02c1  1f3bc3c6-5a29-2e7f-ddda-2ba93a51b8e6  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        run      failed    2020-02-03T13:33:27-05:00  2020-02-03T13:33:30-05:00
60ed109c-3989-9ab6-4922-edbffd600efe  1f3bc3c6-5a29-2e7f-ddda-2ba93a51b8e6  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        run      failed    2020-02-03T13:33:27-05:00  2020-02-03T13:33:30-05:00
6566472b-433d-3d37-f610-0492a0ef187f  1f3bc3c6-5a29-2e7f-ddda-2ba93a51b8e6  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        run      failed    2020-02-03T13:33:27-05:00  2020-02-03T13:33:30-05:00
3a19b38c-e587-3a5c-9bac-c7dc9f80da70  7a78c791-e553-3d4a-f4f1-9426e2b5c3dd  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        stop     failed    2020-02-03T13:32:56-05:00  2020-02-03T13:33:27-05:00
b1a8d187-5edd-688f-e3e2-19b19f5d5e0c  7a78c791-e553-3d4a-f4f1-9426e2b5c3dd  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        stop     failed    2020-02-03T13:32:56-05:00  2020-02-03T13:33:27-05:00
9f00f827-63ba-da46-8156-c9c5413bf1f4  7a78c791-e553-3d4a-f4f1-9426e2b5c3dd  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        stop     failed    2020-02-03T13:32:56-05:00  2020-02-03T13:33:27-05:00
e3396874-555d-5651-81e0-99a4ef74826d  7a78c791-e553-3d4a-f4f1-9426e2b5c3dd  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       1        stop     failed    2020-02-03T13:32:56-05:00  2020-02-03T13:33:27-05:00
3c7eb6cb-aea9-055f-96e6-760eb3526d4a  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        stop     complete  2020-02-03T13:31:22-05:00  2020-02-03T13:33:31-05:00
3f34931c-0e25-bd86-7ca1-1535d945a180  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        stop     complete  2020-02-03T13:31:22-05:00  2020-02-03T13:33:31-05:00
5099bcaf-336b-99ad-40c7-dcb20d39fd47  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        stop     complete  2020-02-03T13:31:22-05:00  2020-02-03T13:33:31-05:00
d48180bc-021e-93c0-0569-9812329a9d21  42de1870-4824-eb91-81c1-5ee7dd22d57d  d6b606f2-7e0d-2a28-14a1-e0e2fe21caec  client1    group       0        stop     complete  2020-02-03T13:31:22-05:00  2020-02-03T13:33:31-05:00
```

after patch, and after the failed deployed reschedule eval runs, the state should look similar to

```
Deployed
Task Group  Promoted  Desired  Canaries  Placed  Healthy  Unhealthy  Progress Deadline
group       false     4        4         8       0        8          2020-02-03T13:47:52-05:00

Allocations
ID                                    Eval ID                               Node ID                               Node Name  Task Group  Version  Desired  Status   Created                    Modified
19d083df-aaab-2d0d-ac92-a903b2668daf  2f952062-236a-19b1-dde0-b87c812b24f4  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       1        run      failed   2020-02-03T13:38:23-05:00  2020-02-03T13:38:26-05:00
3961fd9a-1afb-729f-fe64-7304800f97a3  2f952062-236a-19b1-dde0-b87c812b24f4  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       1        run      failed   2020-02-03T13:38:23-05:00  2020-02-03T13:38:26-05:00
72188056-c001-90df-ef9f-4c69a19401c4  2f952062-236a-19b1-dde0-b87c812b24f4  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       1        run      failed   2020-02-03T13:38:23-05:00  2020-02-03T13:38:26-05:00
fae75b88-6afa-1983-3542-3537b5771a2e  2f952062-236a-19b1-dde0-b87c812b24f4  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       1        run      failed   2020-02-03T13:38:23-05:00  2020-02-03T13:38:26-05:00
de5a77cf-611a-c5a2-2f2a-0826b2301070  b7a73cde-e991-3761-6acd-a063e91c5b73  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       1        stop     failed   2020-02-03T13:37:52-05:00  2020-02-03T13:38:23-05:00
5ee32ced-1dbd-c73d-34ec-34779f2a5769  b7a73cde-e991-3761-6acd-a063e91c5b73  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       1        stop     failed   2020-02-03T13:37:52-05:00  2020-02-03T13:38:23-05:00
7cbee00a-685f-873e-f049-ee3e961082ed  b7a73cde-e991-3761-6acd-a063e91c5b73  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       1        stop     failed   2020-02-03T13:37:52-05:00  2020-02-03T13:38:23-05:00
c12fb7e1-8983-52bd-c1cd-a6e3d4558f73  b7a73cde-e991-3761-6acd-a063e91c5b73  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       1        stop     failed   2020-02-03T13:37:52-05:00  2020-02-03T13:38:23-05:00
38de6ad5-5454-de4d-5a59-0925324d3343  558d9046-f009-1413-5f8c-a72d070eac94  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       0        run      running  2020-02-03T13:37:22-05:00  2020-02-03T13:37:42-05:00
95cbecb4-55f4-4238-e441-33e051369c9f  558d9046-f009-1413-5f8c-a72d070eac94  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       0        run      running  2020-02-03T13:37:22-05:00  2020-02-03T13:37:37-05:00
5aa2911e-bdb3-c8fa-5aba-0670aeae451f  558d9046-f009-1413-5f8c-a72d070eac94  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       0        run      running  2020-02-03T13:37:22-05:00  2020-02-03T13:37:38-05:00
e51c5311-6c70-7c54-5b19-47597494b31c  558d9046-f009-1413-5f8c-a72d070eac94  352beb8d-6906-2997-1cf6-0777e2a37b69  client1    group       0        run      running  2020-02-03T13:37:22-05:00  2020-02-03T13:37:42-05:00

```
